src/main.py

```diff
@@ -634,7 +637,13 @@ def non_ha_replica_iter(self, db_state, zk_state):
                 db_state['wal_receiver'],
                 replics_info,
             )
-            if not streaming and not can_delayed:
+            current_primary = zk_state['lock_holder']
+
+            if streaming_from_primary:
+                self._acquire_replication_source_slot_lock(current_primary)
+            if streaming:
```

Comment:
`elif`? Or some other checks structure, because `current_primary` can be same as `stream_from`, and both `streaming_from_primary` and `streaming` can be true.